### ISSUE TITLE
Phase 59.1: LoRAEngine — Low-Rank Adaptation Module

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#1114]
### Summary
Implement a **LoRAEngine** that performs low-rank adaptation of pre-trained model weights. Instead of updating full weight matrices W ∈ ℝ^{d×k}, LoRA decomposes weight updates as ΔW = BA where B ∈ ℝ^{d×r} and A ∈ ℝ^{r×k} with rank r << min(d,k), achieving parameter efficiency of ~0.1% while maintaining fine-tuning quality.

#### Core LoRA Implementation
- **Low-rank decomposition**: W_new = W_frozen + (α/r) · BA with configurable rank r and scaling α
- **Target module selection**: Automatically identify and wrap attention (Q, K, V, O) and MLP layers
- **Initialization**: Kaiming uniform for A, zero-init for B (ensuring ΔW=0 at start)
- **Rank scheduling**: Dynamic rank allocation — higher ranks for critical layers, lower for redundant ones

#### Advanced LoRA Variants
- **LoRA+**: Differential learning rates for A and B matrices (Hayou et al.)
- **DoRA**: Weight-decomposed LoRA separating magnitude and direction (Liu et al.)
- **rsLoRA**: Rank-stabilized scaling factor √r instead of r for stable training at higher ranks
- **AdaLoRA**: Adaptive rank allocation via SVD-based importance scoring
- **LoRA-FA**: Frozen-A variant — freeze A after initialization, only train B

#### Multi-LoRA Management
- **Adapter stacking**: Merge multiple LoRA adapters for multi-task models
- **LoRA merging strategies**: Linear interpolation, TIES-Merging, DARE for combining adapters
- **Inference optimization**: Merge LoRA weights into base model for zero-overhead inference
- **Hot-swap support**: Load/unload LoRA adapters without model reload

### API Design
```python
class LoRAEngine:
    def apply_lora(self, model, config: LoRAConfig) -> PeftModel
    def select_target_modules(self, model, strategy: str) -> List[str]
    def set_rank_schedule(self, layer_ranks: Dict[str, int]) -> None
    def merge_adapters(self, adapters: List[LoRAAdapter], strategy: str) -> LoRAAdapter
    def merge_and_unload(self, model: PeftModel) -> nn.Module
    def compute_param_efficiency(self, model) -> EfficiencyReport

class LoRAConfig:
    rank: int = 16
    alpha: float = 32.0
    dropout: float = 0.05
    target_modules: List[str] = ["q_proj", "v_proj"]
    variant: str = "standard"  # standard, dora, adalora, lora_plus
    use_rslora: bool = False
```

### Performance Targets
| Metric | Target |
|--------|--------|
| Parameter efficiency | < 1% of base model params |
| Quality retention | > 95% of full fine-tuning |
| Memory reduction | > 60% vs full fine-tuning |
| Adapter merge latency | < 100ms for 7B model |
| Rank selection accuracy | Within 2 ranks of optimal |

### References
- Hu et al. (2022) — LoRA: Low-Rank Adaptation of Large Language Models
- Hayou et al. (2024) — LoRA+: Efficient Low-Rank Adaptation of Large Models
- Liu et al. (2024) — DoRA: Weight-Decomposed Low-Rank Adaptation
- Zhang et al. (2023) — AdaLoRA: Adaptive Budget Allocation for PEFT

### Depends On
- Phase 59.5 (PEFTOrchestrator) for unified pipeline integration

### Part Of
- Phase 59 — Parameter-Efficient Fine-Tuning & Adaptation
- Planning: #1113
